Allow custom implementation of system tag managers

PVince81 opened the pull request:
Added config.php option to replace the default implementation of system
tag manager and system tag object mapper.

Ref: https://github.com/owncloud/core/issues/20258#issue-114788380
Similar to how it was done for comments.

Please review @blizzz @nickvergessen @icewind1991 @DeepDiver1975 @rullzer

Commit:
Allow custom implementation of system tag managers

Added config.php option to replace the default implementation of system
tag manager and system tag object mapper.

Also adjusted the comments manager factory to inject the server container

nickvergessen commented on lib/private/systemtag/managerfactory.php:
```diff
@@ -0,0 +1,61 @@
+	 */
+	public function getManager() {
+		return new SystemTagManager(
+			\OC::$server->getDatabaseConnection()
```
Should still be possible to use DI here

PVince81 replied:
Not really. @blizzz already met this issue.

The problem is that we need the constructor to be generic and have no arguments, because here https://github.com/owncloud/core/pull/21792/files#diff-4bebfeb572bb49e9d3fc4c1c480813d0R150

nickvergessen replied:
Well we could make it generic and inject the server, tada you have DI instead of static/global usage

nickvergessen commented on lib/private/server.php:
```diff
@@ -143,11 +143,18 @@ public function __construct($webRoot, \OC\Config $config) {
 			$tagMapper = $c->query('TagMapper');
 			return new TagManager($tagMapper, $c->getUserSession());
 		});
+		$this->registerService('SystemTagManagerFactory', function (Server $c) {
```
Add to the unit test list:
https://github.com/owncloud/core/blob/master/tests/lib/server.php#L141

Same with the others please

PVince81 replied:
forgot this, will add

PVince81 commented:
- [ ] REQUIRES https://github.com/owncloud/core/pull/21793

PVince81 commented:
I updated the PR.

As a bonus, I also adjusted the comments manager factory to also use DI for the server container.

Please review @nickvergessen @rullzer @blizzz

MorrisJobke commented:
:boom:

```


1) Test\Comments\Test_Comments_Manager::testGetCommentNotFound
Argument 1 passed to OC\Comments\ManagerFactory::__construct() must implement interface OCP\IServerContainer, none given, called in /ssd/jenkins/workspace/core-ci-linux@3/database/sqlite/label/SLAVE/tests/lib/comments/manager.php on line 53 and defined

/ssd/jenkins/workspace/core-ci-linux@3/database/sqlite/label/SLAVE/lib/private/comments/managerfactory.php:41
/ssd/jenkins/workspace/core-ci-linux@3/database/sqlite/label/SLAVE/tests/lib/comments/manager.php:53
/ssd/jenkins/workspace/core-ci-linux@3/database/sqlite/label/SLAVE/tests/lib/comments/manager.php:61

...
```

PVince81 commented:
Right... should have adjusted the comments unit tests too.
Will do.

icewind1991 commented:
:+1: looks good

PVince81 commented:
Fixed the comments manager test.

rullzer commented:
:+1:
